src/main/scala/com/ing/wbaa/airlock/proxy/handler/RequestHandlerS3.scala

```diff
@@ -49,5 +63,102 @@ trait RequestHandlerS3 extends LazyLogging with RadosGatewayHandler {
+      if (isContentsTag) {
+        allContentsElements += element
+        immutable.Seq.empty
+      } else {
+        immutable.Seq(element)
+      }
+    }
+
+    val rangerAuthorizedFalseCache = mutable.Map.empty[String, Boolean]
+
+    def isPathOkInRangerPolicy(path: String): Boolean = {
+      val pathToCheck = normalizePath(path)
+      val isUserAuthorized = isUserAuthorizedForRequest(requestS3.copy(s3BucketPath = Some(pathToCheck)), user)
+      logger.debug("user {} isUserAuthorized for path {} = {}", user.userName, pathToCheck, isUserAuthorized)
+      if (!isUserAuthorized) rangerAuthorizedFalseCache.put(path, isUserAuthorized)
+      isUserAuthorized
+    }
+
+    def normalizePath(path: String): String = {
+      val delimiter = "/"
+      val decodedPath = URLDecoder.decode(path, "UTF-8")
+      val delimiterIndex = decodedPath.lastIndexOf(delimiter)
+      val pathToCheck = if (delimiterIndex > 0) delimiter + decodedPath.substring(0, delimiterIndex) else ""
+      val s3path = requestS3.s3BucketPath.getOrElse(delimiter)
+      val s3pathWithoutLastDelimiter = if (s3path.length > 1 && s3path.endsWith(delimiter)) s3path.substring(0, s3path.length - 1) else s3path
+      s3pathWithoutLastDelimiter + pathToCheck
+    }
+
+    Flow[ByteString].via(XmlParsing.parser)
+      .statefulMapConcat(() => {
+        // state
+        val keyTagValue = StringBuilder.newBuilder
+        val allContentsElements = new ListBuffer[ParseEvent]
+        var isContentsTag = false
+        var isKeyTag = false
+
+        // aggregation function
+        parseEvent =>
+          parseEvent match {
+            case element: StartElement if element.localName == "Contents" =>
+              isContentsTag = true
+              allContentsElements.clear()
+              allContentsElements += element
+              immutable.Seq.empty
+            case element: EndElement if element.localName == "Contents" =>
+              isContentsTag = false
+              allContentsElements += element
+              if (isPathOkInRangerPolicy(keyTagValue.stripMargin)) {
+                allContentsElements.toList
+              } else {
+                immutable.Seq.empty
+              }
+            case element: StartElement if element.localName == "Key" =>
+              keyTagValue.clear()
+              isKeyTag = true
+              elementResult(allContentsElements, isContentsTag, element)
+            case element: EndElement if element.localName == "Key" =>
+              isKeyTag = false
+              elementResult(allContentsElements, isContentsTag, element)
+            case element: TextEvent =>
+              if (isKeyTag) keyTagValue.append(element.text)
+              elementResult(allContentsElements, isContentsTag, element)
+            case s =>
+              elementResult(allContentsElements, isContentsTag, s)
+          }
```

Comment:
Could you add a brief comment on what each case means or a java doc on top of the function? It's quite a large function and not immediately clear to me what it does/how it works

Reply:
done